packages/data/src/plugins/persistence/index.js

```diff
@@ -178,12 +164,25 @@ export default function( registry, pluginOptions ) {
 				return registry.registerStore( reducerKey, options );
 			}
 
-			const initialState = persistence.get()[ reducerKey ];
+			// Load from persistence to use as initial state.
+			let initialState = persistence.get()[ reducerKey ];
+			if ( initialState !== undefined ) {
+				// For object-like persistence, ensure that:
+				// - Other keys are left intact when persisting only a subset
+				//   of keys.
+				// - New keys in what would otherwise be provided as a default
+				//   state are deeply merged as a base for the persisted value.
+				if ( isPlainObject( initialState ) ) {
+					initialState = merge(
+						options.reducer( undefined, {
```

Comment:
I wonder if we should ensure that this also returns a plain object or undefined to avoid a corrupted persisted state from being loaded.

Reply:
What would be the expected behavior if we had a plain-object persisted state and a non-plain-object default initial state?

Reply:
In a1469ff302c696321bd36787beacc2cac84a89b2, I went with: If there's a mismatch of object-likeness between initial and persisted states, it defers always to the value produced by the original reducer.

But immediately upon pushing the change, I'm second-guessing myself, particularly for the following scenarios:

- Why defer to the default implementation for only the mismatch on this type, when if they were both non-objects we'd be fine to use the persisted value, regardless whether they're different types (array vs. string vs. number vs. etc)
- Are there cases where we'd want to allow some "empty" initial state, but still prefer the persisted value? Take, for example, the changed test case "should load a persisted value as initialState" in a1469ff302c696321bd36787beacc2cac84a89b2 where the reducer returns `null` as the default initial state

With consideration of these points, maybe it should be that the persisted value always  "wins", and merges with the default implementation value if and only if both it and the persisted value are objects.

The merging behavior in general feels a bit odd to me, though I understand the need for it in introducing new keys to something like a preferences object over time. The only other ideas that came to mind to support this were things more explicitly defined as "upgrading" behaviors†. There's some overhead in trying to manage this, which I think the auto-merging handles reasonably well on its own. 

† Things like: Handling the "persistence restore" action type from the reducer, checking the value and adding missing keys. Or defining a validation function which returns whether a persisted value should be considered valid for use.

Reply:
Yes, I also was thinking to defer to the reducer value and consider the persisted one as broken but on a second though reducers can return different shapes so I guess it's fine to override the state entirely with the persisted value.

Reply:
> Yes, I also was thinking to defer to the reducer value and consider the persisted one as broken but on a second though reducers can return different shapes so I guess it's fine to override the state entirely with the persisted value.

And should a need arise that we need to explicitly forget some persisted state, I think at that point we could consider introducing one of the other options I mentioned in my previous comment about validating or overriding the persisted input.